libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/BatchingHttpInterceptor.kt

```diff
@@ -66,7 +66,8 @@ class BatchingHttpInterceptor @JvmOverloads constructor(
 ) : HttpInterceptor {
   private val dispatcher = CloseableSingleThreadDispatcher()
   private val scope = CoroutineScope(dispatcher.coroutineDispatcher)
-  private val mutex = Mutex()
+  private val pendingRequestsMutex = Mutex()
+  private val loopMutex = Mutex(locked = true)
```

Comment:
The mutex doesn't need to be locked initially, right? 

Overall, it's a bit weird to use a `Mutex` here, I would have expected a `pthread_cond_t` or some similar MPP equivalent but I guess there's none? 

Also, back to the initial use case of "grouped queries", is this solving the issue?

```
- intercept query1
- unblocks loop immediatly
- query1 gets sent in a batch containing only one query
- intercept query2, query3, query n
- loop is waiting for `batchIntervalMillis`
- query2, query3 and query n are sent batched
```

Reply:
The more I think about this, the more I think your initial hunch of a task to be scheduled when `intercept()` is called was the good one, except of a fixed `batchIntervalMillis` delay, it should keep track of time windows and tune it accordingly?

Reply:
Your example is right :) 

I went back to the scheduling but with an adjusting time window instead of a fixed one, I think the delay of         `batchIntervalMillis - (currentTimeMillis() % batchIntervalMillis) - 1` makes sense but LMK what you think.

Reply:
LGTM 👍